inference/explained_bool.go

```diff
@@ -28,10 +28,12 @@ import (
 // - <Val>BecauseDeepConstraint: Applied to site X when X was half of an assertion where the other half was fixed, but through a deeper chain of assertions
 // - <Val>BecauseAnnotation: Applied to site X when a syntactic annotation was discovered on X
 type ExplainedBool interface {
+	fmt.Stringer
+
 	Val() bool
-	String() string
-	getPrimitiveFullTrigger() primitiveFullTrigger
-	deeperReason() ExplainedBool
+	Pos() token.Pos
+	TriggerReprs() (producer fmt.Stringer, consumer fmt.Stringer)
+	DeeperReason() ExplainedBool
```

Comment:
We have to export these methods since diagnostic handling is in a separate package now. Moreover, it's often bad to have an exported method that returns an unexported struct. 

Since I think we don't want to export `primitiveFullTrigger` (it is only meaningful inside the inference engine), I have exposed these methods instead that are supposed to work across packages.

Reply:
Why not make the type of `producer` and `consumer` as `annotation.Prestring`? Any particular reason we want to use `fmt.Stringer`?

Reply:
`annotation.Prestring` is equivalent to `fmt.Stringer` in that they both expose `String() string` method and can be used interchangeably. So I opted for a standard `fmt.Stringer` instead of `annotation.Prestring`.

I was thinking we could remove `annotation.Prestring` interface definition from our codebase, it is already provided by stdlib and is equally readable (if not more). If we ever need to expand the interface, we should do it [at the client side](https://github.com/golang/go/wiki/CodeReviewComments#interfaces) instead of putting it in the `annotation` library package.

But I don't have strong preferences here and I can revert this to use `annotation.Prestring` if you disagree 😃

Reply:
I see your point. I agree that `fmt.Stringer` and `annotation.Prestring` are the essentially the same interfaces at this point. `fmt.Stringer` is also very less likely to change, since it is used quite ubiquitously, and if we want to expand the interface for our purpose, we can follow the path you are suggesting. Therefore, I am fine using either of them, provided we do so consistently across the entire codebase. 😄 You can keep this change, and perhaps create an issue to track changing of `Prestring` to `Stringer` at other places in the codebase.

Reply:
👍 !

I have created an issue #69 to track this clean-up task!